vendor/tracing/zero/destinations.ml

```diff
@@ -32,7 +32,69 @@ let direct_file_destination ?(buffer_size = 4096 * 16) ~filename () =
   (module Dest : Writer_intf.Destination)
 ;;
 
-let file_destination ~filename () = direct_file_destination ~filename ()
+let compressed_file_destination ?(buffer_size = 4096 * 16) ~filename () =
+  let buf = Iobuf.create ~len:buffer_size in
+  let com_size = buffer_size * 2 in
+  let com_level = 5 in
+  let compressed_buf = Iobuf.create ~len:com_size in
+  let file = Core_unix.openfile ~mode:[ O_CREAT; O_TRUNC; O_RDWR ] filename in
+  let written = ref 0 in
+  let compression_context = Zstandard.Compression_context.create () in
+  let flush () =
+    Iobuf.rewind buf;
+    Iobuf.advance buf !written;
+    Iobuf.flip_lo buf;
+    let input =
+      Zstandard.Input.from_bigstring
+        ~pos:(Iobuf.Expert.lo buf)
+        ~len:(Iobuf.length buf)
+        (Iobuf.Expert.buf buf)
+    in
+    let output =
+      Zstandard.Output.in_buffer
+        ~pos:(Iobuf.Expert.lo compressed_buf)
+        ~len:com_size
+        (Iobuf.Expert.buf compressed_buf)
+    in
+    let compressed_length =
+      Zstandard.With_explicit_context.compress
+        compression_context
+        ~compression_level:com_level
+        ~input
+        ~output
+    in
```

Comment:
We (hlian and I) were uncertain how to reduce code duplication between `compressed_file_destination` and `file_destination`, so we wanted to open this PR up early. We think it might make sense to have a match statement in `let flush` and then _both_ uncompressed and compressed file destinations can use the same function. Something like

```ocaml
(* all the set up *)
let flush () =
  match is_compressed with
  | true -> ...
  | false -> ...
in
(* ... *)
(module Dest : Destination.S)
```

Does this make sense? If so, we can keep pushing code to this branch. Or, perhaps, is there a better way?

Reply:
Ok, leaving it as is!